### DISCUSSION TITLE
Proper way to log things when using DDP

### DISCUSSION THREAD

**Discussion by jandonov** [Lightning-AI/pytorch-lightning#6501] [DDP / multi-GPU / multi-node] [answered]
Hi, I was wondering what is the proper way of logging metrics when using DDP. I noticed that if I want to print something inside `validation_epoch_end` it will be printed twice when using 2 GPUs. I was expecting `validation_epoch_end` to be called only on rank 0 and to receive the outputs from all GPUs, but I am not sure this is correct anymore. Therefore I have several questions:

1. `validation_epoch_end(self, outputs)` - When using DDP does every subprocess receive the data processed from the current GPU or data processed from all GPUs, i.e. does the input parameter `outputs` contains the outputs of the entire validation set, from all GPUs?
2. If `outputs` is GPU/process specific what is the proper way to calculate any metric on the entire validation set in `validation_epoch_end` when using DDP?

I understand that I can solve the printing by checking `self.global_rank == 0` and printing/logging only in that case, however I am trying to get a deeper understanding of what I am printing/logging in this case. 

Here is a code snippet from my use case. I would like to be able to report f1, precision and recall on the entire validation dataset and I am wondering what is the correct way of doing it when using DDP.

    
```Python
    def _process_epoch_outputs(self,
                               outputs: List[Dict[str, Any]]
                               ) -> Tuple[torch.Tensor, torch.Tensor]:
        """Creates and returns tensors containing all labels and predictions

        Goes over the outputs accumulated from every batch, detaches the
        necessary tensors and stacks them together.

        Args:
            outputs (List[Dict])
        """
        all_labels = []
        all_predictions = []

        for output in outputs:
            for labels in output['labels'].detach():
                all_labels.append(labels)

            for predictions in output['predictions'].detach():
                all_predictions.append(predictions)

        all_labels = torch.stack(all_labels).long().cpu()
        all_predictions = torch.stack(all_predictions).cpu()

        return all_predictions, all_labels

    def validation_epoch_end(self, outputs: List[Dict[str, Any]]) -> None:
        """Logs f1, precision and recall on the validation set."""

        if self.global_rank == 0:
            print(f'Validation Epoch: {self.current_epoch}')

        predictions, labels = self._process_epoch_outputs(outputs)
        for i, name in enumerate(self.label_columns):

            f1, prec, recall, t = metrics.get_f1_prec_recall(predictions[:, i],
                                                             labels[:, i],
                                                             threshold=None)
            self.logger.experiment.add_scalar(f'{name}_f1/Val',
                                              f1,
                                              self.current_epoch)
            self.logger.experiment.add_scalar(f'{name}_Precision/Val',
                                              prec,
                                              self.current_epoch)
            self.logger.experiment.add_scalar(f'{name}_Recall/Val',
                                              recall,
                                              self.current_epoch)

            if self.global_rank == 0:
                print((f'F1: {f1}, Precision: {prec}, '
                       f'Recall: {recall}, Threshold {t}'))
```

**Comment by flukeskywalker** [comment]
I have the same question, and have not been able to get sufficient clarity from the docs about how logging works during distributed training. 

I found the suggestion to [use the `sync_dist` flag when logging during validation/testing in distributed training](https://pytorch-lightning.readthedocs.io/en/stable/advanced/multi_gpu.html#synchronize-validation-and-test-logging), but it's unclear exactly what this does, and whether I should or shouldn't use `sync_dist=True` when logging in the training step as well. If not, why not?

**Reply by jandonov** [reply]
@flukeskywalker have you managed to find a solution?

**Comment by rudaoshi** [comment]
I have the same problem.  I managed to log the synced metric by calling metric.compute(). But the value is not identical with the Checkpoint callback.

Details please find in https://github.com/PyTorchLightning/pytorch-lightning/discussions/6352

And there is a related issue: 

https://github.com/PyTorchLightning/pytorch-lightning/issues/6122

**Reply by jandonov** [reply]
@rudaoshi have you solved this problem?

**Reply by rudaoshi** [reply]
No, I just write my own DDP running engine.

**Comment by jandonov** [comment]
@williamFalcon any chance someone can help us with this?

**Reply by williamFalcon** [reply]
cc @edenafek

**Reply by jandonov** [reply]
Ty for pointing us to the right person @williamFalcon!

**Reply by jandonov** [reply]
@williamFalcon Apologies for tagging you again but this seems like a general issue and there are already two more people in this thread that are struggling with the same thing. Unfortunately, @edenafek didn't reply to this thread, so I was wondering if you can point us to a person, place in the documentation where this is explained or a working example that we can use. Will really appreciate it!

**Comment by jandonov** [comment]
@edenafek  could you please take a look at the above issue?

**Comment by SkafteNicki** [comment, marked as answer]
Hi all,
Sorry we have not got back to you in time, let me try to answer some of your questions:
1. Is `validation_epoch_end` only called on rank 0?

No, it is called by all processes

2. What does the `sync_dist` flag do:

Here is the essential code:
https://github.com/PyTorchLightning/pytorch-lightning/blob/a72a7992a283f2eb5183d129a8cf6466903f1dc8/pytorch_lightning/core/step_result.py#L108-L115
If `sync_dist=True` then it will as default call the `sync_ddp` function which will sum the value across all processes using `torch.distributed.all_reduce` 
https://github.com/PyTorchLightning/pytorch-lightning/blob/a72a7992a283f2eb5183d129a8cf6466903f1dc8/pytorch_lightning/utilities/distributed.py#L120
Use this flag if you want to synchronize the value between different processes. 

3. How to print stuff in distributed lightning:

Recommended is using either the `rank_zero_info` function. Import as:
```python
from pytorch_lightning.utilities import rank_zero_info
```
or use the `rank_zero_only` decorator (imported from the same module) which can be wrapped around any function such that it only gets called on `rank=0`. Each logger experiment is decorated with `rank_zero_experiment` which internally calls `rank_zero_only`
https://github.com/PyTorchLightning/pytorch-lightning/blob/a72a7992a283f2eb5183d129a8cf6466903f1dc8/pytorch_lightning/loggers/base.py#L31-L43

4. What about `pytorch_lightning.metrics` (now known as `torchmetrics`)

Our own metrics have custom synchronization going on. Any metric will automatically synchronize between different processes whenever `metric.compute()` is called. Metrics calculated this way should therefore not be logged using `sync_dist=True`.

5. Recommended way of logging:

Using `self.log` in your lightning module 

Not sure this answers all questions.

**Reply by ashleve** [reply]
@SkafteNicki so is there a way to compute something from outputs gathered from all processes? Any possibility of calling
```python
def validation_epoch_end(self, outputs: List[Dict[str, Any]]) -> None:
    ...
```
with all outputs available?

**Reply by SkafteNicki** [reply]
you should be able to call `self.all_gather` on the tensor, to gather results from all processes
https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.core.lightning.html?highlight=gather_all#pytorch_lightning.core.lightning.LightningModule.all_gather

**Reply by jandonov** [reply]
@SkafteNicki So in my particular case in `validation_epoch_end` I first use my `_process_epoch_outputs` function to take out the predictions and the labels from every list item in `outputs` and stack them together in two tensors `predictions` and `labels` that I return.

Does this mean that in my code in `validation_epoch_end` after:

```Python
predictions, labels = self._process_epoch_outputs(outputs)
```

I can simply use something like:

```Python
predictions = self.all_gather(predictions)
labels = self.all_gather(labels)
```
?

If this is true I have several questions:
1. It says in the documentation that the input can be a collection. Does this mean that we can actually call `self.all_gather` on the `outputs` list that we take as an input argument in `validation_epoch_end`?
2. Are there some rules we need to follow in order for `self.all_gather` to be able to locate all corresponding tensors? Like for example not using `.cpu()` or `.detach()` etc.

**Reply by SkafteNicki** [reply]
>     1. It says in the documentation that the input can be a collection. Does this mean that we can actually call `self.all_gather` on the `outputs` list that we take as an input argument in `validation_epoch_end`?

Yes, it should also work with the raw `outputs` tensor. Basically, all tensors in the collection will get a new first dimension equal to the world size (so a tensor of size `[B,d]` will after all gather have size `[world_size, B, d]`)

>     2. Are there some rules we need to follow in order for `self.all_gather` to be able to locate all corresponding tensors? Like for example not using `.cpu()` or `.detach()` etc.

No, it should locate the corresponding tensors on all devices based on variable name. I am pretty sure that you are not allowed to call `.cpu()` but `detach` should be no problem.

**Reply by flukeskywalker** [reply]
@SkafteNicki Thanks for the clarifications! They are very helpful.

It will be wonderful if you can clarify a few more things about logging with `ddp`. Am I understanding correctly that:
1. If one **is** using `torchmetrics` and logging using `self.log()`, then one **never** needs to use `sync_dist=True`? It doesn't matter whether it is a training step or a validation step etc., and both aggregation (average by default) over GPUs and aggregation (average by default) over steps in an epoch is taken care of by Lightning.
2. If one **is not** using `torchmetrics` but just logging some arbitrary scalar of interest (could be any loss or some norm, etc.) using `self.log()`, then one should **always** use `sync_dist=True` to record the correct value (averaged over GPUs) of the scalar, regardless of whether it is a training step/epoch or validation step/epoch. This is very important because these records of custom monitored tensors are what we put in papers.

If 2 above is true, can you please clarify what happens if one forgets to explicitly set `sync_dist=True` for any logged quantity? I'd like to understand what do the resulting logged values actually correspond to.

Apologies for so many questions, and thanks again for your help!

**Reply by SkafteNicki** [reply]
>     1. If one **is** using `torchmetrics` and logging using `self.log()`, then one **never** needs to use `sync_dist=True`? It doesn't matter whether it is a training step or a validation step etc., and both aggregation (average by default) over GPUs and aggregation (average by default) over steps in an epoch is taken care of by Lightning.

Yes, you should never use `torchmetrics` together with `sync_dist=True`, because all syncing are handled whenever `.compute()` is called. The reason are:
If you are logging the metric object directly (which you can, and I recommend doing it) then this if statement makes sures that `sync_dist` have no effect 
https://github.com/PyTorchLightning/pytorch-lightning/blob/22a266d8b8cf57455cc863e20491e416ec635ba7/pytorch_lightning/core/step_result.py#L108
If you are logging the result of `compute` then syncing have already taken place and the result is the same of all devices, meaning that you will be averaging over the same. The value will therefore not be wrong, but it will take extra time.
 
>     2. If one **is not** using `torchmetrics` but just logging some arbitrary scalar of interest (could be any loss or some norm, etc.) using `self.log()`, then one should **always** use `sync_dist=True` to record the correct value (averaged over GPUs) of the scalar, regardless of whether it is a training step/epoch or validation step/epoch. This is very important because these records of custom monitored tensors are what we put in papers.


No, because (and this is maybe my own opinion) nobody cares about the exact value of your training curve. If you see training curve in paper it is often to study the overall shape of the curve not the exact numbers (which are noisy per definition). For validation I would enable `sync_dist=True` since I normally only log metrics calculated over the hole validation set, and not only on a single validation batch (which I do not think make sense since the validation set is intended as a proxy for the generalization error). Additionally, I would never perform testing (which are numbers people actually put in papers) in a distributed environment and in that case `sync_dist` have no effect. 

> If 2 above is true, can you please clarify what happens if one forgets to explicitly set `sync_dist=True` for any logged quantity? I'd like to understand what do the resulting logged values actually correspond to.

You will essentially only get the value recorded on the first process. Again, not a problem for training metrics, can be a problem for validation metrics.

> Apologies for so many questions, and thanks again for your help!

No problem :]

**Reply by flukeskywalker** [reply]
Let's agree to disagree :) 

I prefer to report precise data if I have labeled the y-axis ticks, and if I just want to show the shape I can write that explicitly (and perhaps normalize the y range). Of course in some situations and at some scales this does not make a difference, but this is about 1) good habits and 2) being aware of precisely what my code does. During new model development I am often closely monitoring several properties of the model at every batch, and can miss something if the property is not aggregated over processes. 

In a paper, I think it should be extremely clear what exactly is being plotted in all situations, so now I know that if I don't set `sync_dist=True`, I should say that "we plot the value from the first process".

> You will essentially only get the value recorded on the first process. Again, not a problem for training metrics, can be a problem for validation metrics.

That's what I suspected; thanks for the confirmation!

**Reply by SkafteNicki** [reply]
> Let's agree to disagree :)

I respect your opinion :]
Essentially, is is just a trade-off between the preciseness of the logged numbers and the computational overhead syncing introduces.

**Reply by jandonov** [reply]
@SkafteNicki I will need your expertise again as I can't seem to make `self.all_gather` work :)

I have edited the `validation_epoch_end` function above to:

```Python
def validation_epoch_end(self, outputs: List[Dict[str, Any]]) -> None:
        """Logs f1, precision and recall on the validation set."""

        predictions, labels = self._process_epoch_outputs(outputs)

        if self.global_rank == 0:
            print((f'Global Rank {self.global_rank}, '
                   f'Validation Epoch: {self.current_epoch}'), flush=True)

            print(f'Predictions shape pre all_gather: {predictions.shape}', flush=True)
            print(f'Labels shape pre all_gather: {labels.shape}', flush=True)
            predictions = self.all_gather(predictions)
            labels = self.all_gather(labels)
            print(f'Predictions shape post all_gather: {predictions.shape}', flush=True)
            print(f'Labels shape post all_gather: {labels.shape}', flush=True)
            predictions = predictions.view((-1, self.num_labels))
            labels = labels.view((-1, self.num_labels))
            print(f'Predictions shape post stack: {predictions.shape}', flush=True)
            print(f'Labels shape post stack: {labels.shape}', flush=True)

            thresholds = []
            avg_f1 = 0

            for i, name in enumerate(self.label_columns):

                f1, prec, recall, t = metrics.get_f1_prec_recall(
                    predictions=predictions[:, i].cpu().numpy(),
                    labels=labels[:, i].cpu().numpy(),
                    threshold=None
                )

                thresholds.append(t)
                avg_f1 += f1

                self.logger.experiment.add_scalar(f'{name}_f1/Val',
                                                  f1,
                                                  self.current_epoch)
                self.logger.experiment.add_scalar(f'{name}_Precision/Val',
                                                  prec,
                                                  self.current_epoch)
                self.logger.experiment.add_scalar(f'{name}_Recall/Val',
                                                  recall,
                                                  self.current_epoch)

                print((f'{name} || '
                       f'F1: {f1:.3f}, Precision: {prec:.3f}, '
                       f'Recall: {recall:.3f}, Threshold {t:.3f} '
                       f'Global Rank: {self.global_rank}'))

                avg_f1 /= self.num_labels
                if avg_f1 > self.best_avg_f1:
                    self.best_avg_f1 = avg_f1
                    self.best_thresholds = thresholds
```

The output I get from the print statements is:

```
Global Rank 0, Validation Epoch: 0
Predictions shape pre all_gather: torch.Size([40, 42])
Labels shape pre all_gather: torch.Size([40, 42])
Predictions shape post all_gather: torch.Size([2, 40, 42])
Labels shape post all_gather: torch.Size([2, 40, 42])
Predictions shape post stack: torch.Size([80, 42])
Labels shape post stack: torch.Size([80, 42])
```

I am using 2 GPUs and DDP so at first glance `self.all_gather` seems to work, but then the code simply hangs and doesn't continue with the execution. I can't even stop the execution with CTRL + C and I have to manually kill the process. Additionally, If I comment out the `self.all_gather` lines the code works again without issues.

**Reply by SkafteNicki** [reply]
You need to call `self.all_gather` because the gathering is not done in asynchronous, meaning that all processes will wait for all processes to finish before continue. Since you currently only are executing `self.all_gather` on `rank==0` it will wait forever because `rank==1` will never finish (because it never started).
You really only need to make sure that logging and printing happens on `rank==0`. For logging, note that `self.logger.experiment` is already decorated with `only_rank_zero` decorator:
https://github.com/PyTorchLightning/pytorch-lightning/blob/3baac7125283b94134572549580071e313288016/pytorch_lightning/loggers/tensorboard.py#L129-L131
meaning that logging always only happen on `rank==0` (all other ranks get a dummy experiment that does nothing). For printing, see one of my comments above.

**Reply by jandonov** [reply]
@SkafteNicki thanks a lot, I did experiment in the past week with `self.all_gather` and it seems to work nicely. I was wondering one thing tho: 

How dangerous is it to use it `self.all_gather` on large tensors (e.g. validation/test set predictions) when using multiple GPUs? If I keep my GPUs utilized at nearly 100% memory is it possible that `self.all_gather` will cause an overflow on GPU 0?

**Reply by jiwoosong** [reply]
Can `self.all_gather` bind `list` or `dict` other than tensor?

I am performing 3D segmentation and reconsidering the score for each cluster of predictions. The output is a tensor of different sizes, and I want to collect the results of each distributed processing at rank=0.

Currently, it is synchronized with `torch.distributed.barrier(group)`, and each distributed processing result is saved/loaded with `pickle` to calculate the metric at once. It does not take much time, but errors sometimes occur in the process of reading and writing files. I'd like to know a more stable way.

**Reply by erikdao** [reply]
@jiwoosong I'm dealing with a similar problem as well. Have you figured out a good solution?

**Reply by prateeky2806** [reply]
Thanks for the great question and answer. This helped me a lot.

**Reply by tranngocphu** [reply]
This is the best exchanges by far I found about logging in DDP training. Thank you guys!

**Reply by atnikos** [reply]
Very insightful discussion and lots of nice replies. I want to ask one more thing @SkafteNicki .
I am trying ddp. I set `dist_sync=True` which is supposed to merge values from all GPUs, but I get `pytorch_lightning.utilities.exceptions.MisconfigurationException: You called `self.log(recons/rfeats2rfeats/val, ...)` twice in `validation_epoch_end` with different arguments. This is not allowed.`
When I disable it the code runs fine but from what I read reports only Gpu_0 calculation. Is there any help about this?
My code:

```
# this is just a custom tracker that accumulates results (nn.Module)
loss_tracker = self.tracker[split]
loss_dict = loss_tracker.compute()
loss_tracker.reset()

dico = {loss_tracker.loss2logname(loss, split): value.item()
        for loss, value in loss_dict.items()}

# workaround for LR, assuming 1 optimizer, 1 scheduler, very weak
curr_lr = self.trainer.optimizers[0].param_groups[0]['lr']
dico.update({'Learning Rate': curr_lr})
self.log_dict(dico)

dico.update({"epoch": float(self.trainer.current_epoch),
             "step": float(self.trainer.current_epoch)})

self.log_dict(dico)    # this is the line I change params for dist_sync
```

**Reply by 12michi34** [reply]
@SkafteNicki 
thanx for the answer this helped me with most of the questions I had. 
One thing I am not clear is what @flukeskywalker mentioned as well:
whether I should or shouldn't use sync_dist=True when logging in the training step as well. If not, why not?
https://pytorch-lightning.readthedocs.io/en/1.4.0/advanced/multi_gpu.html mentions only validation and testing so 
I am wondering what I need to do with 
`self.log("train_loos", loss, logger=True)`
in my training_step() function when using DDP. Should I use a TorchMetric or add the sync_dist flag for training as well ?
Cheers

**Reply by Alec-Stashevsky** [reply]
@SkafteNicki I am looking at the code you provide for `self.log` and see that the default sync operation is `"mean"`, not `"sum"`. Can you confirm this is your understanding as well?

**Reply by SkafteNicki** [reply]
Hi @Alec-Stashevsky,
You are completely right that the default option is `mean` and not `sum`, the relevant code line:
https://github.com/Lightning-AI/lightning/blob/d5ca30aaf15aadeab8a1186430e230eb00c8eb4f/src/lightning/pytorch/core/module.py#L356
Cannot remember if `sum` was the original default when this discussion was created two years ago.

**Reply by krunolp** [reply]
@12michi34 I think that the same holds for training, I've been wondering the same and think I found the answer in https://lightning.ai/forums/t/synchronize-train-logging/1270.

**Reply by mfoglio** [reply]
In conclusion, should we use `sync_dist=True` when using ddp (and ddp_find_unused_parameters)?

**Reply by davidgill97** [reply]
IMO, sync_dist is very confusing and dangerous to use, and i think it should be avoided when possible. Using sync_dist means that the metrics are computed at multiple devices and reduced when sync_dist=True, but this has an obvious problem: the reduced metric is not be same as metric computed from aggregated outputs and labels. (E.g. mAP..)


There is a dedicated torchmetrics library allows easier control over aggregation and computation, so i think it is best to use it to calculate metrics, instead of sync_dist. Even for very specific custom metrics, i felt like it's much self-contained and less confusing to implement by inheriting from `torchmetrics.Metric`.
